### PR TITLE
Feat #33/oidc support

### LINKED ISSUE #33: Add support for OIDC
OIDC authentication should be configured as follows:

`client credentials` grant:

* Token is embedded in `Authorization` header
* The OAuth parameters (`client_id`, etc.) should be shown to the user on the OIDC login page

`user password` grant 

* Token is embedded in `Cookie` (e.g. `auth_token=XYZ`)
* The OAuth parameters (`client_id` etc.) should be shown to the user on the OIDC login page

### PR DESCRIPTION
**Pull request by mvhysko**
closes #33

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -39,10 +39,13 @@
     "fastify-session": "^5.2.1",
     "fastify-swagger": "^4.0.1",
     "jose": "^2.0.3",
+    "jsonwebtoken": "^8.5.1",
+    "jwk-to-pem": "^2.0.5",
     "jwt-simple": "^0.5.6",
     "libxmljs": "^0.19.7",
     "node-jwk": "^0.1.0",
     "pg": "^8.5.1",
+    "qs": "^6.10.1",
```
Comment (derevnjuk) [inline package.json:48]: Why don't use the built-in module?

File: src/keycloak/keycloak.service.ts [hunk 2]
```diff
@@ -0,0 +1,222 @@
...
+      ),
+    };
+
+    ok(this.realm, '"realm" is not defined.');
+    ok(this.server_uri, '"metadata_uri" is not defined.');
+    ok(this.clientUser.client_id, '"client_id" is not defined.');
+    ok(this.clientAdmin.client_id, 'Admin "client_id" is not defined.');
+    ok(this.clientAdmin.client_secret, 'Admin "client_secret" is not defined.');
+
+    this.discovery().catch((err: any) => console.error(err));
```
Comment (derevnjuk) [inline src/keycloak/keycloak.service.ts:91]: Why don't use `OnModuleInit` hook or async provider? You should avoid calling long operations from the constructor (async or sync)

File: src/keycloak/keycloak.service.ts [hunk 3]
```diff
@@ -0,0 +1,222 @@
...
+      client_id: this.configService.get(
+        KeyCloakConfigProperties.ENV_KEYCLOAK_ADMIN_CLIENT_ID,
+      ),
+      client_secret: this.configService.get(
+        KeyCloakConfigProperties.ENV_KEYCLOAK_ADMIN_CLIENT_SECRET,
+      ),
+    };
+
+    ok(this.realm, '"realm" is not defined.');
+    ok(this.server_uri, '"metadata_uri" is not defined.');
```
Comment (derevnjuk) [inline src/keycloak/keycloak.service.ts:86]:
```suggestion
    ok(this.server_uri, '"server_uri" is not defined.');
```

File: .gitignore [hunk 4]
```diff
@@ -22,6 +22,7 @@ lerna-debug.log*
 /.nyc_output
 
 # IDEs and editors
+.vscode
```
Comment (derevnjuk) [inline .gitignore:25]: It's already there

**Review by derevnjuk — CHANGES_REQUESTED** [review]

File: .vscode/settings.json [hunk 5]
```diff
@@ -0,0 +1,5 @@
+{
```
Comment (derevnjuk) [inline .vscode/settings.json:1]: Please remove this file and adjust `.gitignore` removing the following lines:

```
!.vscode/settings.json
!.vscode/tasks.json
!.vscode/extensions.json
```

File: public/src/pages/auth/Login/Login.tsx [hunk 6]
```diff
@@ -1,7 +1,13 @@
 import { AxiosRequestConfig } from 'axios';
 import React, { FC, FormEvent, useEffect, useState } from 'react';
 import { Link } from 'react-router-dom';
-import { getLdap, getUser, loadXsrfToken } from '../../../api/httpClient';
+import { OidcClient, OidcClientType } from 'src/interfaces/Auth';
```
Comment (derevnjuk) [inline public/src/pages/auth/Login/Login.tsx:4]: Use only relative imports or TS paths

File: src/auth/auth.controller.ts [hunk 7]
```diff
@@ -182,6 +185,20 @@ export class AuthController {
     return token;
   }
 
+  @Get('oidc-client')
+  async getOidcClient(
+    @Query() query: OidcClientRequest,
+  ): Promise<OidcClientResponse> {
+    const client = this.keyCloakService.getClient(
+      (query.type as ClientType) || ClientType.ADMIN,
+    );
```
Comment (derevnjuk) [inline src/auth/auth.controller.ts:194]: You must never return private "ADMIN" secrets, you should keep them safe and secure.

File: src/keycloak/keycloak.service.ts [hunk 8]
```diff
@@ -39,13 +40,18 @@ export interface Token {
   readonly expires_in?: number;
 }
 
+export enum ClientType {
+  ADMIN = 'admin',
+  USER = 'user',
```
Comment (derevnjuk) [inline src/keycloak/keycloak.service.ts:45]: I would prefer to use a more clear and convenient name

```suggestion
  PUBLIC = 'public',
```

**Review by derevnjuk — CHANGES_REQUESTED** [review]
Great job, @mvhysko 💪🏾 Please fix the remaining issues and reassign the ticket to @rielas

File: src/auth/auth.controller.ts (hunk 7 above)
Comment (mvhysko) [reply src/auth/auth.controller.ts:194]: so, I should return only clientId?

**Comment by derevnjuk** [comment]
@mvhysko please resolve conflicts and ensure that signup via `keycloak` is failure-tolerant and idempotent

File: docker-compose.yml [hunk 9]
```diff
@@ -47,5 +50,37 @@ services:
...
+      DB_ADDR: keycloak-db
+      DB_DATABASE: keycloak
+      DB_USER: keycloak
+      DB_SCHEMA: public
+      DB_PASSWORD: password
+      KEYCLOAK_USER: admin
+      KEYCLOAK_PASSWORD: Pa55w0rd
+      KEYCLOAK_IMPORT: /opt/jboss/keycloak/imports/realm-export.json -Dkeycloak.profile.feature.upload_scripts=enabled
+    ports:
+      - "8080:8080"
```
Comment (derevnjuk) [inline docker-compose.yml:81]:
```suggestion
```

**Review by derevnjuk — APPROVED** [review]
@mvhysko please resolve conflicts and ensure that signup via `keycloak` is failure-tolerant and idempotent

File: src/keycloak/keycloak.service.ts [hunk 10]
```diff
@@ -0,0 +1,238 @@
...
+        ...tokenData,
+        grant_type: 'password',
+      });
+    } else {
+      Object.assign(tokenPayload, {
+        grant_type: 'client_credentials',
+      });
+    }
+
+    try {
```
Comment (derevnjuk) [inline src/keycloak/keycloak.service.ts:192]: The caller code already logs errors, try/catch block is redundant

File: src/keycloak/keycloak.service.ts [hunk 11]
```diff
@@ -0,0 +1,238 @@
...
+    firstName,
+    lastName,
+    email,
+    password,
+  }: RegisterUserData): Promise<void> {
+    this.log.debug(`Called registerUser`);
+
+    const { access_token, token_type } = await this.generateToken();
+
+    try {
```
Comment (derevnjuk) [inline src/keycloak/keycloak.service.ts:145]: ditto

**Review by derevnjuk — APPROVED** [review]

**Review by derevnjuk — APPROVED** [review]

**Pull request merged**